zaza/openstack/charm_tests/octavia/tests.py

```diff
@@ -99,11 +100,77 @@ def revoke_role_current_user(keystone_client, keystone_session, role_name,
 class CharmOperationTest(test_utils.OpenStackBaseTest):
     """Charm operation tests."""
 
+    application_name = "octavia"
+    states = {"octavia": {
+        "workload-status": "active",
+        "workload-status-message": "Unit is ready"}}
+
     @classmethod
     def setUpClass(cls):
         """Run class setup for running Octavia charm operation tests."""
         super(CharmOperationTest, cls).setUpClass()
 
+    def get_leaders_and_non_leaders(self):
+        """Get leader node and non-leader nodes of octavia."""
+        status = zaza.model.get_status().applications[self.application_name]
+        self.leader = None
+        self.non_leaders = []
+        for unit in status["units"]:
+            if status["units"][unit].get("leader"):
+                self.leader = unit
+            else:
+                self.non_leaders.append(unit)
+        return self.leader, self.non_leaders
+
+    def get_port_ips(self):
+        """Extract IP info from Neutron ports tagged with charm-octavia."""
+        keystone_session = openstack_utils.get_overcloud_keystone_session()
+        neutron_client = openstack_utils.get_neutron_session_client(
+            keystone_session)
+        resp = neutron_client.list_ports(tags='charm-octavia')
+        neutron_ip_list = []
+        for port in resp['ports']:
+            # one new DOWN port may be created when deleting one octavia unit
+            if port['status'] == 'DOWN':
+                continue
+            for ip_info in port['fixed_ips']:
+                neutron_ip_list.append(ip_info['ip_address'])
+        return neutron_ip_list
+
+    def test_update_controller_ip_port_list(self):
+        """Test update_controller_ip_port_list.
+
+        Remove a unit, and then query the list of ports to check that the port
+        has been deleted.
+        """
+        logging.info("test_update_controller_ip_port_list: start test")
+        logging.info("Wait till model is idle ...")
+        zaza.model.block_until_all_units_idle()
+        ips = self.get_port_ips()
+        num = len(ips)
+        logging.info('initial hm port num is {}: {}'.format(num, ips))
+
+        logging.info("test_update_controller_ip_port_list: add one unit")
+        logging.info("Adding one unit ...")
+        zaza.model.add_unit(self.application_name)
+        logging.info("Wait until one unit is added ...")
+        zaza.model.block_until_unit_count(self.application_name, num+1)
+        zaza.model.wait_for_application_states(states=self.states)
```

Comment:
This is the same as `zaza.model.wait_for_application_states()`, so you don't need `self.states` at all. This is a bit lengthy to explain here, but in a nutshell, both will lead to "waiting for **all** applications to be active/ready"

Reply:
done